File: packages/cli/src/commands/list.ts
import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import path from "node:path";

import { resolveRunxWorkspaceBase } from "@runxhq/core/config";
import {
  parseRunnerManifestYaml,
  parseToolManifestJson,
  validateRunnerManifest,
  validateToolManifest,
} from "@runxhq/core/parser";

import {
  buildLocalPacketIndex,
  countYamlFiles,
  discoverSkillProfilePaths,
  safeReadDir,
  toProjectPath,
} from "../authoring-utils.js";

export type RunxListRequestedKind = "all" | "tools" | "skills" | "chains" | "packets" | "overlays";
export type RunxListItemKind = Exclude<RunxListRequestedKind, "all"> extends infer Kind
  ? Kind extends string
    ? Kind extends `${infer Singular}s`
      ? Singular
      : Kind
    : never
  : never;
export type RunxListSource = "local" | "workspace" | "dependencies" | "built-in";

export interface RunxListItem {
  readonly kind: RunxListItemKind;
  readonly name: string;
  readonly source: RunxListSource;
  readonly path: string;
  readonly status: "ok" | "invalid";
  readonly diagnostics?: readonly string[];
  readonly scopes?: readonly string[];
  readonly emits?: readonly { readonly name: string; readonly packet?: string }[];
  readonly fixtures?: number;
  readonly harness_cases?: number;
  readonly steps?: number;
  readonly wraps?: string;
}

export interface RunxListReport {
  readonly schema: "runx.list.v1";
  readonly root: string;
  readonly requested_kind: RunxListRequestedKind;
  readonly items: readonly RunxListItem[];
}

export interface ListCommandArgs {
  readonly listKind?: RunxListRequestedKind;
  readonly listOkOnly?: boolean;
  readonly listInvalidOnly?: boolean;
}

export async function handleListCommand(parsed: ListCommandArgs, env: NodeJS.ProcessEnv): Promise<RunxListReport> {
  const root = resolveRunxWorkspaceBase(env);
  const requestedKind = parsed.listKind ?? "all";
  const items = await discoverListItems(root, requestedKind);
  const filtered = items.filter((item) => {
    if (parsed.listOkOnly) {
      return item.status === "ok";
    }
    if (parsed.listInvalidOnly) {
      return item.status === "invalid";
    }
    return true;
  });
  return {
    schema: "runx.list.v1",
    root,
    requested_kind: requestedKind,
    items: sortListItems(filtered),
  };
}

export function normalizeListKind(value: string | undefined): RunxListRequestedKind | undefined {
  if (value === undefined || value === "") {
    return "all";
  }
  if (["tools", "skills", "chains", "packets", "overlays"].includes(value)) {
    return value as RunxListRequestedKind;
  }
  return undefined;
}

async function discoverListItems(root: string, requestedKind: RunxListRequestedKind): Promise<readonly RunxListItem[]> {
  const items: RunxListItem[] = [];
  if (requestedKind === "all" || requestedKind === "tools") {
    items.push(...await discoverToolListItems(root));
  }
  if (requestedKind === "all" || requestedKind === "skills" || requestedKind === "chains") {
    items.push(...(await discoverSkillAndChainListItems(root)).filter((item) => requestedKind === "all" || `${item.kind}s` === requestedKind));
  }
  if (requestedKind === "all" || requestedKind === "packets") {
    items.push(...await discoverPacketListItems(root));
  }
  if (requestedKind === "all" || requestedKind === "overlays") {
    items.push(...await discoverOverlayListItems(root));
  }
  return items;
}

async function discoverToolListItems(root: string): Promise<readonly RunxListItem[]> {
  const toolsRoot = path.join(root, "tools");
  const items: RunxListItem[] = [];
  for (const namespaceEntry of await safeReadDir(toolsRoot)) {
    if (!namespaceEntry.isDirectory()) {
      continue;
    }
    const namespaceDir = path.join(toolsRoot, namespaceEntry.name);
    for (const toolEntry of await safeReadDir(namespaceDir)) {
      if (!toolEntry.isDirectory()) {
        continue;
      }
      const manifestPath = path.join(namespaceDir, toolEntry.name, "manifest.json");
      if (!existsSync(manifestPath)) {
        continue;
      }
      const relativePath = toProjectPath(root, manifestPath);
      try {
        const tool = validateToolManifest(parseToolManifestJson(await readFile(manifestPath, "utf8")));
        const emits = tool.artifacts?.namedEmits
          ? Object.entries(tool.artifacts.namedEmits).map(([name, packet]) => ({ name, packet }))
          : tool.artifacts?.wrapAs
            ? [{ name: tool.artifacts.wrapAs }]
            : [];
        items.push({
          kind: "tool",
          name: tool.name,
          source: "local",
          path: relativePath,
          status: "ok",
          scopes: tool.scopes,
          emits,
          fixtures: await countYamlFiles(path.join(namespaceDir, toolEntry.name, "fixtures")),
        });
      } catch {
        items.push({
          kind: "tool",
          name: `${namespaceEntry.name}.${toolEntry.name}`,
          source: "local",
          path: relativePath,
          status: "invalid",
          diagnostics: ["runx.tool.manifest.invalid"],
        });
      }
    }
  }
  return items;
}

async function discoverSkillAndChainListItems(root: string): Promise<readonly RunxListItem[]> {
  const items: RunxListItem[] = [];
  for (const profilePath of await discoverSkillProfilePaths(root)) {
    const skillDir = path.dirname(profilePath);
    const fallbackName = skillDir === root ? path.basename(root) : path.basename(skillDir);
    const relativePath = toProjectPath(root, profilePath);
    try {
      const manifest = validateRunnerManifest(parseRunnerManifestYaml(await readFile(profilePath, "utf8")));
      const runners = Object.values(manifest.runners);
      const chainSteps = runners
        .map((runner) => runner.source.chain?.steps.length)
        .filter((value): value is number => typeof value === "number");
      const isChain = chainSteps.length > 0;
      items.push({
        kind: isChain ? "chain" : "skill",
        name: manifest.skill ?? fallbackName,
        source: "local",
        path: relativePath,
        status: "ok",
        fixtures: await countYamlFiles(path.join(skillDir, "fixtures")),
        harness_cases: manifest.harness?.cases.length ?? 0,
        steps: isChain ? chainSteps.reduce((sum, value) => sum + value, 0) : undefined,
      });
    } catch {
      items.push({
        kind: "skill",
        name: fallbackName,
        source: "local",
        path: relativePath,
        status: "invalid",
        diagnostics: ["runx.skill.profile.invalid"],
      });
    }
  }
  return items;
}

async function discoverPacketListItems(root: string): Promise<readonly RunxListItem[]> {
  const index = await buildLocalPacketIndex(root, { writeCache: false });
  return [
    ...index.packets.map((packet) => ({
      kind: "packet" as const,
      name: packet.id,
      source: "local" as const,
      path: packet.path,
      status: "ok" as const,
    })),
    ...index.errors.map((error) => ({
      kind: "packet" as const,
      name: error.ref,
      source: "local" as const,
      path: error.path,
      status: "invalid" as const,
      diagnostics: [error.id],
    })),
  ];
}

async function discoverOverlayListItems(root: string): Promise<readonly RunxListItem[]> {
  const overlaysRoot = path.join(root, "skills-overlays");
  const items: RunxListItem[] = [];
  for (const vendorEntry of await safeReadDir(overlaysRoot)) {
    if (!vendorEntry.isDirectory()) {
      continue;
    }
    const vendorDir = path.join(overlaysRoot, vendorEntry.name);
    for (const skillEntry of await safeReadDir(vendorDir)) {
      if (!skillEntry.isDirectory()) {
        continue;
      }
      const profilePath = path.join(vendorDir, skillEntry.name, "X.yaml");
      if (!existsSync(profilePath)) {
        continue;
      }
      const contents = await readFile(profilePath, "utf8");
      const wraps = /^\s*wraps:\s*(.+?)\s*$/m.exec(contents)?.[1];
      items.push({
        kind: "overlay",
        name: `${vendorEntry.name}/${skillEntry.name}`,
        source: "local",
        path: toProjectPath(root, profilePath),
        status: "ok",
        wraps,
      });
    }
  }
  return items;
}

function sortListItems(items: readonly RunxListItem[]): readonly RunxListItem[] {
  const tierOrder: Record<RunxListSource, number> = {
    local: 0,
    workspace: 1,
    dependencies: 2,
    "built-in": 3,
  };
  const kindOrder: Record<RunxListItemKind, number> = {
    tool: 0,
    skill: 1,
    chain: 2,
    packet: 3,
    overlay: 4,
  };
  return [...items].sort((left, right) =>
    tierOrder[left.source] - tierOrder[right.source]
    || kindOrder[left.kind] - kindOrder[right.kind]
    || left.name.localeCompare(right.name)
  );
}

File: crates/runx-runtime/src/list.rs
// rust-style-allow: large-file because native list discovery intentionally keeps
// tool, skill, graph, packet, and overlay projection in one audited cutover
// surface until the TypeScript list command is fully retired.
use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::{Path, PathBuf};

pub use runx_contracts::{
    RunxListEmit, RunxListItem, RunxListItemKind, RunxListReport, RunxListRequestedKind,
    RunxListSchema, RunxListSource, RunxListStatus,
};
use serde::Deserialize;

use crate::RuntimeError;
use crate::filesystem::{find_files_named, read_dir_sorted, read_to_string};
use crate::path_util::{count_yaml_files, display_path, lexical_normalize, project_path};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RunxListOptions {
    pub root: PathBuf,
    pub requested_kind: RunxListRequestedKind,
}

#[must_use]
pub fn default_list_options(root: PathBuf) -> RunxListOptions {
    RunxListOptions {
        root,
        requested_kind: RunxListRequestedKind::All,
    }
}

pub fn list_authoring_primitives(
    options: &RunxListOptions,
) -> Result<RunxListReport, RuntimeError> {
    let root = lexical_normalize(&options.root);
    let mut items = discover_list_items(&root, options.requested_kind)?;
    sort_list_items(&mut items);
    Ok(RunxListReport {
        schema: RunxListSchema::V1,
        root: display_path(&root),
        requested_kind: options.requested_kind,
        items,
    })
}

fn discover_list_items(
    root: &Path,
    requested_kind: RunxListRequestedKind,
) -> Result<Vec<RunxListItem>, RuntimeError> {
    let mut items = Vec::new();
    if matches!(
        requested_kind,
        RunxListRequestedKind::All | RunxListRequestedKind::Tools
    ) {
        items.extend(discover_tool_list_items(root)?);
    }
    if matches!(
        requested_kind,
        RunxListRequestedKind::All | RunxListRequestedKind::Skills | RunxListRequestedKind::Graphs
    ) {
        items.extend(
            discover_skill_and_graph_list_items(root)?
                .into_iter()
                .filter(|item| match requested_kind {
                    RunxListRequestedKind::All => true,
                    RunxListRequestedKind::Skills => {
                        matches!(item.kind, RunxListItemKind::Skill | RunxListItemKind::Graph)
                    }
                    RunxListRequestedKind::Graphs => item.kind == RunxListItemKind::Graph,
                    _ => false,
                }),
        );
    }
    if matches!(
        requested_kind,
        RunxListRequestedKind::All | RunxListRequestedKind::Packets
    ) {
        items.extend(discover_packet_list_items(root)?);
    }
    if matches!(
        requested_kind,
        RunxListRequestedKind::All | RunxListRequestedKind::Overlays
    ) {
        items.extend(discover_overlay_list_items(root)?);
    }
    Ok(items)
}

fn discover_tool_list_items(root: &Path) -> Result<Vec<RunxListItem>, RuntimeError> {
    let tools_root = root.join("tools");
    let mut items = Vec::new();
    for namespace_entry in read_dir_sorted(&tools_root)? {
        if !namespace_entry.is_dir {
            continue;
        }
        for tool_entry in read_dir_sorted(&namespace_entry.path)? {
            if !tool_entry.is_dir {
                continue;
            }
            let manifest_path = tool_entry.path.join("manifest.json");
            if !manifest_path.exists() {
                continue;
            }
            let relative_path = project_path(root, &manifest_path);
            match read_validated_tool_manifest(&manifest_path) {
                Ok(tool) => items.push(RunxListItem {
                    kind: RunxListItemKind::Tool,
                    name: tool.name,
                    source: RunxListSource::Local,
                    path: relative_path,
                    status: RunxListStatus::Ok,
                    diagnostics: None,
                    scopes: Some(tool.scopes),
                    emits: tool
                        .artifacts
                        .as_ref()
                        .map(tool_emits)
                        .filter(|items| !items.is_empty()),
                    fixtures: Some(count_yaml_files(&tool_entry.path.join("fixtures"))?),
                    harness_cases: None,
                    steps: None,
                    wraps: None,
                }),
                Err(()) => items.push(invalid_item(
                    RunxListItemKind::Tool,
                    format!("{}.{}", namespace_entry.name, tool_entry.name),
                    relative_path,
                    "runx.tool.manifest.invalid",
                )),
            }
        }
    }
    Ok(items)
}

fn read_validated_tool_manifest(manifest_path: &Path) -> Result<runx_parser::ValidatedTool, ()> {
    let source = fs::read_to_string(manifest_path).map_err(|_| ())?;
    let raw = runx_parser::parse_tool_manifest_json(&source).map_err(|_| ())?;
    runx_parser::validate_tool_manifest(raw).map_err(|_| ())
}

fn tool_emits(artifacts: &runx_parser::SkillArtifactContract) -> Vec<RunxListEmit> {
    if let Some(named_emits) = &artifacts.named_emits {
        return named_emits
            .keys()
            .map(|name| RunxListEmit {
                name: name.clone(),
                packet: artifacts
                    .packets
                    .as_ref()
                    .and_then(|packets| packets.get(name))
                    .cloned(),
            })
            .collect();
    }
    artifacts
        .wrap_as
        .iter()
        .map(|name| RunxListEmit {
            name: name.clone(),
            packet: artifacts.packet.clone(),
        })
        .collect()
}

fn discover_skill_and_graph_list_items(root: &Path) -> Result<Vec<RunxListItem>, RuntimeError> {
    let mut items = Vec::new();
    for profile_path in discover_skill_profile_paths(root)? {
        let skill_dir = profile_path.parent().map_or(root, |parent| parent);
        let fallback_name = fallback_skill_name(root, skill_dir);
        let relative_path = project_path(root, &profile_path);
        match read_validated_runner_manifest(&profile_path) {
            Ok(manifest) => {
                let graph_steps = manifest
                    .runners
                    .values()
                    .filter_map(|runner| {
                        runner
                            .source
                            .graph
                            .as_ref()
                            .map(|graph| graph.steps.len() as u64)
                    })
                    .collect::<Vec<_>>();
                let is_graph = !graph_steps.is_empty();
                let scopes = skill_scopes(&manifest);
                let emits = skill_emits(&manifest);
                items.push(RunxListItem {
                    kind: if is_graph {
                        RunxListItemKind::Graph
                    } else {
                        RunxListItemKind::Skill
                    },
                    name: manifest.skill.unwrap_or(fallback_name),
                    source: RunxListSource::Local,
                    path: relative_path,
                    status: RunxListStatus::Ok,
                    diagnostics: None,
                    scopes,
                    emits,
                    fixtures: Some(count_yaml_files(&skill_dir.join("fixtures"))?),
                    harness_cases: Some(
                        manifest
                            .harness
                            .as_ref()
                            .map_or(0, |harness| harness.cases.len() as u64),
                    ),
                    steps: is_graph.then(|| graph_steps.iter().sum()),
                    wraps: None,
                });
            }
            Err(()) => items.push(invalid_item(
                RunxListItemKind::Skill,
                fallback_name,
                relative_path,
                "runx.skill.profile.invalid",
            )),
        }
    }
    Ok(items)
}

fn skill_scopes(manifest: &runx_parser::SkillRunnerManifest) -> Option<Vec<String>> {
    let mut scopes = BTreeSet::new();
    for runner in manifest.runners.values() {
        if let Some(values) = runner
            .raw
            .get("scopes")
            .and_then(runx_contracts::JsonValue::as_array)
        {
            scopes.extend(
                values
                    .iter()
                    .filter_map(runx_contracts::JsonValue::as_str)
                    .map(str::to_owned),
            );
        }
        if let Some(graph) = &runner.source.graph {
            for step in &graph.steps {
                scopes.extend(step.scopes.iter().cloned());
            }
        }
    }
    (!scopes.is_empty()).then(|| scopes.into_iter().collect())
}

fn skill_emits(manifest: &runx_parser::SkillRunnerManifest) -> Option<Vec<RunxListEmit>> {
    let mut emits = BTreeSet::<(String, Option<String>)>::new();
    for runner in manifest.runners.values() {
        if let Some(artifacts) = &runner.artifacts {
            emits.extend(
                tool_emits(artifacts)
                    .into_iter()
                    .map(|emit| (emit.name, emit.packet)),
            );
        }
        if let Some(graph) = &runner.source.graph {
            for step in &graph.steps {
                if let Some(artifacts) = step.artifacts.as_ref() {
                    emits.extend(json_artifact_emits(artifacts));
                }
            }
        }
    }
    (!emits.is_empty()).then(|| {
        emits
            .into_iter()
            .map(|(name, packet)| RunxListEmit { name, packet })
            .collect()
    })
}

fn json_artifact_emits(artifacts: &runx_contracts::JsonObject) -> Vec<(String, Option<String>)> {
    if let Some(named) = artifacts
        .get("named_emits")
        .and_then(runx_contracts::JsonValue::as_object)
    {
        let packets = artifacts
            .get("packets")
            .and_then(runx_contracts::JsonValue::as_object);
        return named
            .keys()
            .map(|name| {
                let packet = packets
                    .and_then(|packets| packets.get(name))
                    .and_then(runx_contracts::JsonValue::as_str)
                    .map(str::to_owned);
                (name.clone(), packet)
            })
            .collect();
    }
    artifacts
        .get("wrap_as")
        .and_then(runx_contracts::JsonValue::as_str)
        .map(|name| {
            let packet = artifacts
                .get("packet")
                .and_then(runx_contracts::JsonValue::as_str)
                .map(str::to_owned);
            vec![(name.to_owned(), packet)]
        })
        .unwrap_or_default()
}

fn read_validated_runner_manifest(
    profile_path: &Path,
) -> Result<runx_parser::SkillRunnerManifest, ()> {
    let source = fs::read_to_string(profile_path).map_err(|_| ())?;
    let raw = runx_parser::parse_runner_manifest_yaml(&source).map_err(|_| ())?;
    runx_parser::validate_runner_manifest(raw).map_err(|_| ())
}

// rust-style-allow: long-function because packet discovery keeps glob expansion,
// schema-id extraction, and duplicate-id diagnostics in one deterministic pass.
fn discover_packet_list_items(root: &Path) -> Result<Vec<RunxListItem>, RuntimeError> {
    let package_json_path = root.join("package.json");
    if !package_json_path.exists() {
        return Ok(Vec::new());
    }

    let source = read_to_string(&package_json_path)?;
    let package_json = match serde_json::from_str::<PackageJson>(&source) {
        Ok(package_json) => package_json,
        Err(_) => {
            return Ok(vec![invalid_item(
                RunxListItemKind::Packet,
                "package.json".to_owned(),
                "package.json".to_owned(),
                "runx.packet.package.invalid",
            )]);
        }
    };

    let mut items = Vec::new();
    let mut seen = BTreeMap::<String, String>::new();
    for packet_glob in package_json
        .runx
        .as_ref()
        .map(|runx| runx.packets.as_slice())
        .unwrap_or_default()
    {
        let files = expand_local_glob(root, packet_glob)?;
        if files.is_empty() {
            items.push(invalid_item(
                RunxListItemKind::Packet,
                packet_glob.clone(),
                "package.json".to_owned(),
                "runx.packet.ref.missing",
            ));
            continue;
        }
        for file_path in files {
            let relative_path = project_path(root, &file_path);
            let source = match fs::read_to_string(&file_path) {
                Ok(source) => source,
                Err(_) => {
                    items.push(invalid_item(
                        RunxListItemKind::Packet,
                        relative_path.clone(),
                        relative_path,
                        "runx.packet.schema.invalid",
                    ));
                    continue;
                }
            };
            let schema = match serde_json::from_str::<PacketSchema>(&source) {
                Ok(schema) => schema,
                _ => {
                    items.push(invalid_item(
                        RunxListItemKind::Packet,
                        relative_path.clone(),
                        relative_path,
                        "runx.packet.schema.invalid",
                    ));
                    continue;
                }
            };
            let Some(packet_id) = packet_id(&schema) else {
                items.push(invalid_item(
                    RunxListItemKind::Packet,
                    relative_path.clone(),
                    relative_path,
                    "runx.packet.id.mismatch",
                ));
                continue;
            };
            if let Some(existing_source) = seen.get(&packet_id) {
                if existing_source != &source {
                    items.push(invalid_item(
                        RunxListItemKind::Packet,
                        packet_id,
                        relative_path,
                        "runx.packet.id.collision",
                    ));
                    continue;
                }
            }
            seen.insert(packet_id.clone(), source);
            items.push(RunxListItem {
                kind: RunxListItemKind::Packet,
                name: packet_id,
                source: RunxListSource::Local,
                path: relative_path,
                status: RunxListStatus::Ok,
                diagnostics: None,
                scopes: None,
                emits: None,
                fixtures: None,
                harness_cases: None,
                steps: None,
                wraps: None,
            });
        }
    }
    Ok(items)
}

fn discover_overlay_list_items(root: &Path) -> Result<Vec<RunxListItem>, RuntimeError> {
    let overlays_root = root.join("skills-overlays");
    let mut items = Vec::new();
    for vendor_entry in read_dir_sorted(&overlays_root)? {
        if !vendor_entry.is_dir {
            continue;
        }
        for skill_entry in read_dir_sorted(&vendor_entry.path)? {
            if !skill_entry.is_dir {
                continue;
            }
            let profile_path = skill_entry.path.join("X.yaml");
            if !profile_path.exists() {
                continue;
            }
            let contents = read_to_string(&profile_path)?;
            items.push(RunxListItem {
                kind: RunxListItemKind::Overlay,
                name: format!("{}/{}", vendor_entry.name, skill_entry.name),
                source: RunxListSource::Local,
                path: project_path(root, &profile_path),
                status: RunxListStatus::Ok,
                diagnostics: None,
                scopes: None,
                emits: None,
                fixtures: None,
                harness_cases: None,
                steps: None,
                wraps: overlay_wraps(&contents),
            });
        }
    }
    Ok(items)
}

fn invalid_item(
    kind: RunxListItemKind,
    name: String,
    path: String,
    diagnostic: &str,
) -> RunxListItem {
    RunxListItem {
        kind,
        name,
        source: RunxListSource::Local,
        path,
        status: RunxListStatus::Invalid,
        diagnostics: Some(vec![diagnostic.to_owned()]),
        scopes: None,
        emits: None,
        fixtures: None,
        harness_cases: None,
        steps: None,
        wraps: None,
    }
}

#[derive(Deserialize)]
struct PackageJson {
    runx: Option<PackageRunxConfig>,
}

#[derive(Deserialize)]
struct PackageRunxConfig {
    #[serde(default)]
    packets: Vec<String>,
}

#[derive(Deserialize)]
struct PacketSchema {
    #[serde(rename = "x-runx-packet-id")]
    packet_id: Option<String>,
    #[serde(rename = "$id")]
    schema_id: Option<String>,
}

fn packet_id(schema: &PacketSchema) -> Option<String> {
    schema
        .packet_id
        .as_deref()
        .or(schema.schema_id.as_deref())
        .map(str::to_owned)
}

fn expand_local_glob(root: &Path, glob: &str) -> Result<Vec<PathBuf>, RuntimeError> {
    if !glob.contains('*') {
        let path = root.join(glob);
        return Ok(path.exists().then_some(path).into_iter().collect());
    }

    let normalized = glob.replace('\\', "/");
    let Some(star) = normalized.find('*') else {
        return Ok(Vec::new());
    };
    let base = &normalized[..star];
    let base_dir = base.rfind('/').map_or("", |slash| &base[..=slash]);
    let suffix = &normalized[star + 1..];
    let mut files = read_dir_sorted(&root.join(base_dir))?
        .into_iter()
        .filter(|entry| entry.is_file && display_path(&entry.path).ends_with(suffix))
        .map(|entry| entry.path)
        .collect::<Vec<_>>();
    files.sort();
    Ok(files)
}

fn discover_skill_profile_paths(root: &Path) -> Result<Vec<PathBuf>, RuntimeError> {
    let mut paths = Vec::new();
    let root_profile = root.join("X.yaml");
    if root_profile.exists() {
        paths.push(root_profile);
    }
    paths.extend(find_files_named(&root.join("skills"), "X.yaml")?);
    paths.sort();
    Ok(paths)
}

fn fallback_skill_name(root: &Path, skill_dir: &Path) -> String {
    if skill_dir == root {
        return root.file_name().map_or_else(
            || ".".to_owned(),
            |name| name.to_string_lossy().into_owned(),
        );
    }
    skill_dir.file_name().map_or_else(
        || ".".to_owned(),
        |name| name.to_string_lossy().into_owned(),
    )
}

fn overlay_wraps(contents: &str) -> Option<String> {
    contents.lines().find_map(|line| {
        let trimmed = line.trim();
        trimmed
            .strip_prefix("wraps:")
            .map(str::trim)
            .filter(|value| !value.is_empty())
            .map(str::to_owned)
    })
}

fn sort_list_items(items: &mut [RunxListItem]) {
    items.sort_by(|left, right| {
        source_order(left.source)
            .cmp(&source_order(right.source))
            .then_with(|| kind_order(left.kind).cmp(&kind_order(right.kind)))
            .then_with(|| left.name.cmp(&right.name))
    });
}

fn source_order(source: RunxListSource) -> u8 {
    match source {
        RunxListSource::Local => 0,
        RunxListSource::Workspace => 1,
        RunxListSource::Dependencies => 2,
        RunxListSource::BuiltIn => 3,
    }
}

fn kind_order(kind: RunxListItemKind) -> u8 {
    match kind {
        RunxListItemKind::Tool => 0,
        RunxListItemKind::Skill => 1,
        RunxListItemKind::Graph => 2,
        RunxListItemKind::Packet => 3,
        RunxListItemKind::Overlay => 4,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn overlay_wraps_reads_plain_wraps_line() {
        assert_eq!(
            overlay_wraps("name: demo\n  wraps: vendor/base\n"),
            Some("vendor/base".to_owned())
        );
    }

    #[test]
    fn named_emit_lists_only_explicit_packet_binding() {
        let artifacts = runx_parser::SkillArtifactContract {
            emits: None,
            named_emits: Some(BTreeMap::from([("plan".to_owned(), "plan".to_owned())])),
            packets: Some(BTreeMap::from([(
                "plan".to_owned(),
                "runx.plan.v1".to_owned(),
            )])),
            wrap_as: None,
            packet: None,
        };

        assert_eq!(
            tool_emits(&artifacts),
            vec![RunxListEmit {
                name: "plan".to_owned(),
                packet: Some("runx.plan.v1".to_owned()),
            }]
        );
    }

    #[test]
    fn named_emit_output_name_is_not_reported_as_a_packet() {
        let artifacts = runx_parser::SkillArtifactContract {
            emits: None,
            named_emits: Some(BTreeMap::from([("plan".to_owned(), "plan".to_owned())])),
            packets: None,
            wrap_as: None,
            packet: None,
        };

        assert_eq!(
            tool_emits(&artifacts),
            vec![RunxListEmit {
                name: "plan".to_owned(),
                packet: None,
            }]
        );
    }

    #[test]
    fn sorts_by_kind_then_name() {
        let mut items = vec![
            valid_item(RunxListItemKind::Packet, "b"),
            valid_item(RunxListItemKind::Tool, "z"),
            valid_item(RunxListItemKind::Tool, "a"),
            valid_item(RunxListItemKind::Skill, "a"),
        ];
        sort_list_items(&mut items);
        assert_eq!(
            items
                .iter()
                .map(|item| (item.kind, item.name.as_str()))
                .collect::<Vec<_>>(),
            vec![
                (RunxListItemKind::Tool, "a"),
                (RunxListItemKind::Tool, "z"),
                (RunxListItemKind::Skill, "a"),
                (RunxListItemKind::Packet, "b"),
            ]
        );
    }

    fn valid_item(kind: RunxListItemKind, name: &str) -> RunxListItem {
        RunxListItem {
            kind,
            name: name.to_owned(),
            source: RunxListSource::Local,
            path: ".".to_owned(),
            status: RunxListStatus::Ok,
            diagnostics: None,
            scopes: None,
            emits: None,
            fixtures: None,
            harness_cases: None,
            steps: None,
            wraps: None,
        }
    }
}
